src/EditorFeatures/Core/EmbeddedLanguages/Json/JsonEmbeddedBraceMatcher.cs

```diff
@@ -0,0 +1,110 @@
+        {
+            _info = info;
+        }
+
+        public async Task<BraceMatchingResult?> FindBracesAsync(
+            Document document, int position, BraceMatchingOptions options, CancellationToken cancellationToken)
+        {
+            if (!options.HighlightRelatedJsonComponentsUnderCursor)
+                return null;
+
+            var root = await document.GetRequiredSyntaxRootAsync(cancellationToken).ConfigureAwait(false);
+            var token = root.FindToken(position);
+
+            var semanticModel = await document.GetRequiredSemanticModelAsync(cancellationToken).ConfigureAwait(false);
+
+            var detector = JsonLanguageDetector.GetOrCreate(semanticModel.Compilation, _info);
+            var tree = detector.TryParseString(token, semanticModel, cancellationToken);
+            if (tree == null)
+                return null;
+
+            return GetMatchingBraces(tree, position);
+        }
+
+        private static BraceMatchingResult? GetMatchingBraces(JsonTree tree, int position)
+        {
+            var virtualChar = tree.Text.FirstOrNull(vc => vc.Span.Contains(position));
+            if (virtualChar == null)
+                return null;
+
+            var ch = virtualChar.Value;
+            return ch.Value is '{' or '[' or '(' or '}' or ']' or ')'
+                ? FindBraceHighlights(tree, ch)
+                : null;
+        }
+
+        private static BraceMatchingResult? FindBraceHighlights(JsonTree tree, VirtualChar ch)
+        {
+            var node = FindObjectOrArrayNode(tree.Root, ch);
+            return node switch
+            {
+                JsonObjectNode obj => Create(obj.OpenBraceToken, obj.CloseBraceToken),
+                JsonArrayNode array => Create(array.OpenBracketToken, array.CloseBracketToken),
+                JsonConstructorNode cons => Create(cons.OpenParenToken, cons.CloseParenToken),
+                _ => null,
+            };
+        }
+
+        private static BraceMatchingResult? Create(JsonToken open, JsonToken close)
+        {
+            return open.IsMissing || close.IsMissing
+                ? null
+                : new BraceMatchingResult(open.GetSpan(), close.GetSpan());
+        }
+
+        private static JsonValueNode? FindObjectOrArrayNode(JsonNode node, VirtualChar ch)
+        {
+            switch (node)
+            {
+                case JsonArrayNode array when Matches(array.OpenBracketToken, array.CloseBracketToken, ch):
+                    return array;
+
+                case JsonObjectNode obj when Matches(obj.OpenBraceToken, obj.CloseBraceToken, ch):
+                    return obj;
+
+                case JsonConstructorNode cons when Matches(cons.OpenParenToken, cons.CloseParenToken, ch):
+                    return cons;
+            }
+
+            foreach (var child in node)
+            {
+                if (child.IsNode)
+                {
+                    var result = FindObjectOrArrayNode(child.Node, ch);
```

Comment:
Did you want to do a span check to avoid recursing across the entire JSON and instead recurse where you'll best find it?

Reply:
that makes sense :)

Reply:
@CyrusNajmabadi Did this happen or no? (it's possible it snuck into some other place somewhere?)